x/distribution/types/store.go

```diff
@@ -0,0 +1,35 @@
+package types
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// historical rewards for a validator
+// TODO add reference counter, ref https://github.com/cosmos/cosmos-sdk/pull/3099#discussion_r245747051
+// height is implicit within the store key
+type ValidatorHistoricalRewards = sdk.DecCoins
+
+// current rewards for a validator
+type ValidatorCurrentRewards struct {
+	Rewards sdk.DecCoins `json:"rewards"` // current rewards
+	Period  uint64       `json:"period"`  // current period
+}
+
+// accumulated commission for a validator
+type ValidatorAccumulatedCommission = sdk.DecCoins
+
+// starting info for a delegator reward period
+type DelegatorStartingInfo struct {
+	PreviousPeriod uint64  `json:"previous_period"` // period at which the delegation should withdraw starting from
+	Stake          sdk.Dec `json:"stake"`           // amount of stake delegated
```

Comment:
It's higher precision, since delegations with shares (initially one stake) to a validator which gets slashed will be worth some fraction of one stake each - so by storing it as `sdk.Dec` we can calculate rewards more accurately.

We could change it to `sdk.Int` and round down - that's not the end of the world, although it would result in slightly less precision.

Updated the comment.

Reply:
Oh I understand... yes... yes... this kinda sucks. So even though a validator is slashed to a rounded atom, the delegators within the validator may be left with less than a full token. Yeah I think let's leave it, however, I'm adding this explanation to the comment on this struct.